File: backend/app/services/llm_gateway.py
"""
LLM Gateway Service - Handles all AI/LLM API calls
"""
import httpx
from typing import List, Dict, Optional, AsyncIterator
from app.config import settings

class LLMGateway:
    def __init__(self):
        self.gateway_url = settings.VERCEL_AI_GATEWAY_URL
        self.api_key = settings.VERCEL_AI_GATEWAY_API_KEY
        self.default_temperature = 0.3
    
    async def call(
        self,
        messages: List[Dict[str, str]],
        model: str,
        temperature: Optional[float] = None
    ) -> str:
        """
        Call the LLM gateway with messages.
        Returns the assistant's response text.
        """
        if not self.gateway_url or not self.api_key:
            raise ValueError("Gateway env vars missing")
        
        if not model:
            raise ValueError("Model is required")
        
        model_to_use = model
        temperature_to_use = temperature if temperature is not None else self.default_temperature
        
        # Clean up gateway URL (remove trailing slashes)
        url = self.gateway_url.rstrip('/') + '/chat/completions'
        
        async with httpx.AsyncClient() as client:
            response = await client.post(
                url,
                headers={
                    'content-type': 'application/json',
                    'authorization': f'Bearer {self.api_key}',
                    'x-vercel-ai-key': self.api_key
                },
                json={
                    'model': model_to_use,
                    'temperature': temperature_to_use,
                    'messages': messages
                },
                timeout=120.0  # 2 minute timeout for long generations
            )
            
            if not response.is_success:
                error_data = response.json() if response.headers.get('content-type', '').startswith('application/json') else {}
                error_msg = error_data.get('error', {}).get('message') or error_data.get('message') or f'LLM HTTP {response.status_code}'
                raise Exception(error_msg)
            
            data = response.json()
            return str(data.get('choices', [{}])[0].get('message', {}).get('content', ''))
    
    async def stream(
        self,
        messages: List[Dict[str, str]],
        model: str,
        temperature: Optional[float] = None
    ) -> AsyncIterator[str]:
        """
        Stream the LLM gateway response.
        Yields chunks of text as they become available.
        """
        if not self.gateway_url or not self.api_key:
            raise ValueError("Gateway env vars missing")
        
        if not model:
            raise ValueError("Model is required")
        
        model_to_use = model
        temperature_to_use = temperature if temperature is not None else self.default_temperature
        
        # Clean up gateway URL (remove trailing slashes)
        url = self.gateway_url.rstrip('/') + '/chat/completions'
        
        async with httpx.AsyncClient() as client:
            async with client.stream(
                'POST',
                url,
                headers={
                    'content-type': 'application/json',
                    'authorization': f'Bearer {self.api_key}',
                    'x-vercel-ai-key': self.api_key
                },
                json={
                    'model': model_to_use,
                    'temperature': temperature_to_use,
                    'messages': messages,
                    'stream': True
                },
                timeout=120.0
            ) as response:
                if not response.is_success:
                    error_data = await response.aread()
                    try:
                        import json
                        error_json = json.loads(error_data.decode())
                        error_msg = error_json.get('error', {}).get('message') or error_json.get('message') or f'LLM HTTP {response.status_code}'
                    except:
                        error_msg = f'LLM HTTP {response.status_code}'
                    raise Exception(error_msg)
                
                chunk_count = 0
                async for line in response.aiter_lines():
                    if not line.strip():
                        continue
                    
                    # Handle SSE format: data: {...}
                    if line.startswith('data: '):
                        line = line[6:]  # Remove 'data: ' prefix
                    
                    if line.strip() == '[DONE]' or line.strip() == 'data: [DONE]':
                        print(f"LLM Gateway: Received [DONE] signal after {chunk_count} chunks")
                        break
                    
                    try:
                        import json
                        chunk_data = json.loads(line)
                        
                        # Handle different response formats
                        # OpenAI format: choices[0].delta.content
                        # Some gateways might use different structure
                        content = ''
                        if 'choices' in chunk_data:
                            delta = chunk_data.get('choices', [{}])[0].get('delta', {})
                            content = delta.get('content', '')
                        elif 'content' in chunk_data:
                            content = chunk_data.get('content', '')
                        
                        if content:
                            chunk_count += 1
                            yield content
                    except json.JSONDecodeError as e:
                        # Log but continue - might be incomplete JSON
                        print(f"LLM Gateway: JSON decode error: {e}, line: {line[:100]}")
                        continue
                    except Exception as e:
                        # Log unexpected errors but continue
                        print(f"LLM Gateway: Unexpected error: {e}")
                        continue
                
                print(f"LLM Gateway: Stream completed. Total chunks: {chunk_count}")

File: frontend/src/lib/server/services/llmGateway.ts
import { createLogger, errorMessage } from '@/lib/server/logging';

const GATEWAY_URL = process.env.VERCEL_AI_GATEWAY_URL || '';
const GATEWAY_API_KEY = process.env.VERCEL_AI_GATEWAY_API_KEY || '';

// Single, fixed values for all environments to keep behavior predictable
const DEFAULT_TIMEOUT_MS = 60000;
const MAX_RETRIES = 2;
const RETRY_BASE_DELAY_MS = 1000;
const log = createLogger('llm.gateway', {
  label: 'LLM Gateway',
  eventLabels: {
    initialized: 'Gateway Initialized',
    call_start: 'Call Started',
    api_retry: 'API Retry Scheduled',
    usage: 'Token Usage',
    network_retry: 'Network Retry Scheduled',
    call_failed: 'Call Failed',
  },
});

export type Message = { role: 'system' | 'user' | 'assistant'; content: string };

export class LLMGateway {
	private url: string;
	private apiKey: string;
	private defaultTemperature = 0.3;

	constructor() {
		this.url = GATEWAY_URL.replace(/\/+$/, '');
		this.apiKey = GATEWAY_API_KEY;

		if (!this.url || !this.apiKey) {
			throw new Error('LLM gateway configuration is missing. Please check VERCEL_AI_GATEWAY_URL and VERCEL_AI_GATEWAY_API_KEY environment variables.');
		}
    log.debug('initialized', { gatewayConfigured: true });
	}

  async call(
    messages: Message[],
    model: string,
    temperature?: number,
    abortSignal?: AbortSignal,
    options: { timeoutMs?: number; maxRetries?: number; retryDelayMs?: number } = {}
  ): Promise<string> {

    const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    const maxRetries = Math.max(0, options.maxRetries ?? MAX_RETRIES);
    const retryDelayMs = options.retryDelayMs ?? RETRY_BASE_DELAY_MS;

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      const isLast = attempt === maxRetries;
      try {
        log.debug('call_start', {
          model,
          attempt: attempt + 1,
          maxAttempts: maxRetries + 1,
          timeoutMs,
        });
        // Use provided AbortSignal or create one for timeout
        let controller: AbortController;
        let timeoutId: NodeJS.Timeout | undefined;

        if (abortSignal) {
          controller = { signal: abortSignal } as AbortController;
        } else {
          controller = new AbortController();
          timeoutId = setTimeout(() => controller.abort(), timeoutMs);
        }

        const response = await fetch(`${this.url}/chat/completions`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            authorization: `Bearer ${this.apiKey}`,
            'x-vercel-ai-key': this.apiKey,
          },
          body: JSON.stringify({
            model,
            temperature: temperature ?? this.defaultTemperature,
            messages,
          }),
          signal: controller.signal,
        });

        if (timeoutId) clearTimeout(timeoutId);

        const payload = await response.json().catch(() => ({}));
        if (!response.ok) {
          const message = payload?.error?.message || payload?.message || `HTTP ${response.status}`;
          const retryable = response.status === 429 || response.status >= 500;
          if (!isLast && retryable) {
            const delay = retryDelayMs * Math.pow(2, attempt);
            log.warn('api_retry', {
              model,
              attempt: attempt + 1,
              maxAttempts: maxRetries + 1,
              status: response.status,
              retryInMs: delay,
              reason: message,
            });
            await new Promise((r) => setTimeout(r, delay));
            continue;
          }
          throw new Error(message);
        }

        const usage = payload?.usage || {};
        log.debug('usage', {
          model,
          attempt: attempt + 1,
          promptTokens: usage.prompt_tokens ?? null,
          completionTokens: usage.completion_tokens ?? null,
          totalTokens: usage.total_tokens ?? null,
        });

        const content = payload?.choices?.[0]?.message?.content;
        return typeof content === 'string' ? content.trim() : '';
      } catch (error: unknown) {
        const isAbort = error instanceof Error && error.name === 'AbortError';
        const message = isAbort ? `timeout after ${timeoutMs}ms` : (error instanceof Error ? error.message : 'Unknown network error');
        const retryable = isAbort || (error instanceof Error && /ECONNRESET|ETIMEDOUT|ENOTFOUND|fetch failed/i.test(error.message));

        if (!isLast && retryable) {
          const delay = retryDelayMs * Math.pow(2, attempt);
          log.warn('network_retry', {
            model,
            attempt: attempt + 1,
            maxAttempts: maxRetries + 1,
            retryInMs: delay,
            reason: message,
          });
          await new Promise((r) => setTimeout(r, delay));
          continue;
        }

        log.error('call_failed', {
          model,
          attempt: attempt + 1,
          maxAttempts: maxRetries + 1,
          reason: errorMessage(error),
          timeoutMs,
        });

        throw new Error(isAbort ? `LLM API call timed out after ${timeoutMs}ms` : `LLM API call failed: ${message}`);
      }
    }

    throw new Error('LLM call failed after retries');
  }

	async *stream(messages: Message[], model: string, temperature?: number) {
		const response = await fetch(`${this.url}/chat/completions`, {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
				authorization: `Bearer ${this.apiKey}`,
				'x-vercel-ai-key': this.apiKey,
			},
			body: JSON.stringify({
				model,
				temperature: temperature ?? this.defaultTemperature,
				messages,
				stream: true,
			}),
		});

		if (!response.ok || !response.body) {
			const payload = await response.json().catch(() => ({}));
			throw new Error(payload?.error?.message || payload?.message || `LLM HTTP ${response.status}`);
		}

		const reader = response.body.getReader();
		const decoder = new TextDecoder();

		try {
			while (true) {
				const { done, value } = await reader.read();
				if (done) break;

				const chunk = decoder.decode(value, { stream: true });

				for (const line of chunk.split('\n')) {
					const trimmed = line.trim();
					if (!trimmed || trimmed === 'data: [DONE]') continue;

					const dataLine = trimmed.startsWith('data: ')
						? trimmed.slice(6)
						: trimmed;

					try {
						const parsed = JSON.parse(dataLine);
						const delta = parsed?.choices?.[0]?.delta?.content;
						if (delta) {
							yield delta;
						}
					} catch {
						continue;
					}
				}
			}
		} finally {
			reader.releaseLock();
		}
	}
}
